### PR TITLE
Internal: Props schema constraints [EDS-372]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 1]
```diff
@@ -135,7 +128,38 @@ final public function get_atomic_settings(): array {
...
+			if ( ! $prop->validate( $prop->get_default() ) ) {
+				Utils::safe_throw( "Default value for `$key` prop is not of type `{$prop->get_type()}` in `{$widget_name}`." );
+			}
+
+			// Validate default value against the constraints.
+			foreach ( $prop->get_constraints() as $constraint ) {
+				if ( ! $constraint->validate( $prop->get_default() ) ) {
+					Utils::safe_throw( "Default value for `$key` prop does not pass the constraint `{$constraint->get_type()}` in `{$widget_name}`." );
+				}
+			}
```
Comment (eyalnathan) [inline modules/atomic-widgets/base/atomic-widget-base.php:161]: move to validate_default()?

File: modules/atomic-widgets/schema/atomic-prop.php [hunk 2]
```diff
@@ -21,8 +35,86 @@ public function get_default() {
...
+	 */
+	public function number( array $constraints = [] ): self {
+		return $this->type( 'number', $constraints );
+	}
+
+	/**
+	 * @param array<Prop_Constraint> $constraints
+	 */
+	public function boolean( array $constraints = [] ): self {
+		return $this->type( 'boolean', $constraints );
```
Comment (eyalnathan) [inline modules/atomic-widgets/schema/atomic-prop.php:69]: extract type names to consts

File: modules/atomic-widgets/base/atomic-widget-base.php (hunk 1 above)
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:161]: done

File: modules/atomic-widgets/schema/atomic-prop.php (hunk 2 above)
Comment (StyleShit) [reply modules/atomic-widgets/schema/atomic-prop.php:69]: good idea, done

File: tests/phpunit/elementor/modules/atomic-widgets/schema/constraints/test-enum.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+use Elementor\Modules\AtomicWidgets\Schema\Constraints\Enum;
+use ElementorEditorTesting\Elementor_Test_Base;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Test_Enum extends Elementor_Test_Base {
+
+	public function test_construct__throws_when_not_all_values_are_strings() {
```
Comment (Omerisra6) [inline tests/phpunit/elementor/modules/atomic-widgets/schema/constraints/test-enum.php:14]: Happy path

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php [hunk 4]
```diff
@@ -185,12 +195,100 @@ public function test_get_atomic_settings__skip_the_value_transformation_when_it_
...
+
+		// Expect.
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( 'Default value for `prop_with_wrong_default_type` prop is not of type `string`' );
+
+		// Act.
+		$widget::get_props_schema();
+	}
+
+	public function test_get_props_schema__throws_when_default_value_doesnt_pass_constraints() {
```
Comment (Omerisra6) [inline tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:248]:
```suggestion
	public function test_get_props_schema__throws_when_default_value_doesnt_pass_constraints_validation() {
```

**Review by Omerisra6 — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/schema/atomic-prop.php [hunk 5]
```diff
@@ -21,8 +40,92 @@ public function get_default() {
...
+	}
+
+	/**
+	 * @return array<Prop_Constraint>
+	 */
+	public function get_constraints(): array {
+		return $this->constraints;
+	}
+
+	// TODO: Move to a `Schema_Validator` class?
```
Comment (Omerisra6) [inline modules/atomic-widgets/schema/atomic-prop.php:88]: I think it is a better approach, but maybe not in this PR?

**Review by Omerisra6 — DISMISSED** [review]

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 6]
```diff
@@ -43,19 +44,19 @@ private function get_valid_controls( array $controls ): array {
...
 
 			if ( ! $prop_name ) {
-				$this->safe_throw( 'Control is missing a bound prop from the schema.' );
+				Utils::safe_throw( 'Control is missing a bound prop from the schema.' );
 				continue;
 			}
 
 			if ( ! array_key_exists( $prop_name, $schema ) ) {
-				$this->safe_throw( "Prop `{$prop_name}` is not defined in the schema of `{$this->get_name()}`. Did you forget to define it?" );
+				Utils::safe_throw( "Prop `{$prop_name}` is not defined in the schema of `{$this->get_name()}`. Did you forget to define it?" );
```
Comment (ManorHazaz) [inline modules/atomic-widgets/base/atomic-widget-base.php:59]:
```suggestion
				Utils::safe_throw( "Prop `{$prop_name}` is not defined in the schema of `{$this->get_name()}`." );
```

File: modules/atomic-widgets/widgets/atomic-heading.php [hunk 7]
```diff
@@ -89,12 +90,18 @@ protected function define_atomic_controls(): array {
...
 		return [
-			'classes' => Atomic_Prop::make(),
+			'classes' => Atomic_Prop::make()
+				->type( 'classes' )
+				->default( [] ),
 
 			'tag' => Atomic_Prop::make()
+				->string( [
+					Enum::make( [ 'h1', 'h2', 'h3', 'h4', 'h5', 'h6' ] ),
+				] )
```
Comment (ManorHazaz) [inline modules/atomic-widgets/widgets/atomic-heading.php:100]: Maybe?
```suggestion
				->string()
				->constraints( [
					Enum::make( [ 'h1', 'h2', 'h3', 'h4', 'h5', 'h6' ] ),
				] )
```

File: modules/atomic-widgets/schema/atomic-prop.php [hunk 8]
```diff
@@ -21,8 +40,92 @@ public function get_default() {
...
+				return is_bool( $value );
+
+			default:
+				// TODO: Use a shared `is_transformable` function.
+				// TODO: Check the `$value['value']` validity against the relevant transformer.
+				return $this->is_transformable_of_type( $this->get_type(), $value );
+		}
+	}
+
+	private function is_transformable_of_type( string $type, $value ): bool {
```
Comment (ManorHazaz) [inline modules/atomic-widgets/schema/atomic-prop.php:107]: Not sure about the naming here

**Review by ManorHazaz — CHANGES_REQUESTED** [review]

File: includes/utils.php [hunk 9]
```diff
@@ -882,4 +882,12 @@ public static function is_sale_time(): bool {
 
 		return $now_time >= $sale_start_time && $now_time <= $sale_end_time;
 	}
+
+	public static function safe_throw( string $message ) {
```
Comment (blackcolt) [inline includes/utils.php:886]: maybe a function is elementor debug?

File: modules/atomic-widgets/schema/constraints/enum.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+		}
+
+		$this->allowed_values = $allowed_values;
+	}
+
+	public static function make( array $allowed_values ): self {
+		return new static( $allowed_values );
+	}
+
+	public function validate( $value ): void {
```
Comment (blackcolt) [inline modules/atomic-widgets/schema/constraints/enum.php:32]: Do we need to return something from this?

File: modules/atomic-widgets/schema/constraints/enum.php [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+		if ( ! $is_allowed ) {
+			$values = Collection::make( $this->allowed_values )
+				->map( fn ( $value ) => "`$value`" )
+				->implode( ', ' );
+
+			throw new \Exception( "`$value` is not in the list of allowed values ($values)." );
+		}
+	}
+
+	private function all_are_strings( array $allowed_values ): bool {
```
Comment (blackcolt) [inline modules/atomic-widgets/schema/constraints/enum.php:48]: best name ever.

File: modules/atomic-widgets/base/atomic-widget-base.php (hunk 6 above)
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:59]: Done

File: includes/utils.php (hunk 9 above)
Comment (StyleShit) [reply includes/utils.php:886]: Done

File: modules/atomic-widgets/schema/constraints/enum.php (hunk 10 above)
Comment (StyleShit) [reply modules/atomic-widgets/schema/constraints/enum.php:32]: no

File: modules/atomic-widgets/schema/constraints/enum.php (hunk 11 above)
Comment (StyleShit) [reply modules/atomic-widgets/schema/constraints/enum.php:48]: yes

File: modules/atomic-widgets/widgets/atomic-heading.php (hunk 7 above)
Comment (StyleShit) [reply modules/atomic-widgets/widgets/atomic-heading.php]: Done

**Review by ManorHazaz — APPROVED** [review]
Wonderful, Great Job, Amazing

**Pull request merged**